
<!DOCTYPE html>

<html lang="English">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Welcome to the BTE documentation! &#8212; BTE v0.6.0-alpha documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
          

          <div class="body" role="main">
            
  <section id="welcome-to-the-bte-documentation">
<h1>Welcome to the BTE documentation!<a class="headerlink" href="#welcome-to-the-bte-documentation" title="Permalink to this headline">¶</a></h1>
<p>BTE (Big Tree Explorer) is a Python extension for analysis and traversal of extremely large phylogenetic trees. It's based on the
Mutation Annotated Tree (MAT) library which underlies UShER and matUtils, premier software for pandemic-scale phylogenetics.</p>
<p>This tool is generally intended as a replacement for ETE3, Biopython.Phylo, and similar Python phylogenetics packages for Mutation Annotated Trees (MATs).
Using standard packages with MATs requires conversion to newick and the maintenance of mutation annotations as a separate data structure,
generally causing inconvenience and slowing both development and runtime. BTE streamlines this process by exposing the heavily optimized MAT library underlying UShER and matUtils to Python, allowing for efficient and convenient use of MATs in a Python development environment!</p>
<p>UCSC maintains a repository, updated each day, containing the complete and latest publicly-available global SARS-CoV-2 phylogenetic tree in MAT protobuf format here.</p>
<p><a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/">http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/</a></p>
<p>To try out this tool, download the latest tree, build or install the extension, and jump straight to your Python analysis!</p>
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>On Mac or Linux, you can install BTE by running the following command:</p>
<blockquote>
<div><p>conda install -c jmcbroome bte</p>
</div></blockquote>
<p>See the README if you encounter installation difficulties or need to build a local version of the extension.</p>
</section>
<section id="basic-logic">
<h1>Basic Logic<a class="headerlink" href="#basic-logic" title="Permalink to this headline">¶</a></h1>
<p>BTE contains two primary classes, the MATree and the MATNode.</p>
<p>The MATree class represents a single mutation-annotated phylogenetic tree. It can be created from a MAT protobuf (.pb) file, from a newick and a vcf together, or from an Auspice-format JSON like that used by Nextstrain.
The MATree class has many useful functions for manipulation, subsetting, traversal, and summarization of the tree it represents.</p>
<p>The MATNode class represents a single node member of the Mutation Annotated Tree. It has parent, child, mutation, and annotation attributes
which can be used to traverse or compute statistics from the tree.</p>
<p>The full API documentation follows.</p>
</section>
<section id="bte-api-documentation">
<h1>BTE API Documentation<a class="headerlink" href="#bte-api-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p>
<span class="target" id="module-bte"></span><dl class="py class">
<dt class="sig sig-object py" id="bte.MATNode">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bte.</span></span><span class="sig-name descname"><span class="pre">MATNode</span></span><a class="headerlink" href="#bte.MATNode" title="Permalink to this definition">¶</a></dt>
<dd><p>A wrapper around the MAT node class. Has an identifier, mutations, parent, and child attributes.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bte.MATNode.is_leaf">
<span class="sig-name descname"><span class="pre">is_leaf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATNode.is_leaf" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATNode.update_mutations">
<span class="sig-name descname"><span class="pre">update_mutations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATNode.update_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a list of mutations as strings and replace any currently stored mutations on this branch with the new set.
Mutation strings should be formatted as chro:reflocalt e.g. chr1:A234G. If chromosome is left off, assumes SARS-CoV-2 chromosome.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bte.MATree">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bte.</span></span><span class="sig-name descname"><span class="pre">MATree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pb_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncondense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nwk_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nwk_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">json_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree" title="Permalink to this definition">¶</a></dt>
<dd><p>A wrapper around the MAT Tree class. Includes functions to save and load from parsimony .pb files or a newick. Includes 
numerous functions for tree traversal including breadth-first, depth-first, and traversal from leaf to roots. Also includes
numerous functions for subtree selection by choosing leaves that match regex patterns, contain specific mutations, or 
are from a specific clade or lineage.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.apply_mutations">
<span class="sig-name descname"><span class="pre">apply_mutations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mmap</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.apply_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Pass a dictionary of node:mutation list mappings (e.g. {&quot;node_id&quot;:[&quot;chro:reflocalt&quot;,&quot;chro:reflocalt&quot;]}, {&quot;node_1&quot;:[&quot;chro1:A123G&quot;,&quot;chro3:T315G&quot;]})
to place onto the tree. Replaces any mutations currently stored in the indicated nodes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.breadth_first_expansion">
<span class="sig-name descname"><span class="pre">breadth_first_expansion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.breadth_first_expansion" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a level order (breadth-first) expansion starting from the indicated node. Use reverse to traverse in reverse level order (all leaves, then all leaf parents, back to root) instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Call this function to explicitly deallocate all tree memory. Use when the tree object is no longer necessary 
and high memory use is becoming problematic. Automatically called on garbage collection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.compute_nucleotide_diversity">
<span class="sig-name descname"><span class="pre">compute_nucleotide_diversity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.compute_nucleotide_diversity" title="Permalink to this definition">¶</a></dt>
<dd><p>Function which computes the nucleotide diversity of the tree.
This is defined as the mean number of pairwise differences in nucleotides between any two leaves
of the tree. Computes an unbiased estimator which multiplies the final mean by the total number of sequences divided by the total number of sequences minus one.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.count_clades_inclusive">
<span class="sig-name descname"><span class="pre">count_clades_inclusive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subroot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.count_clades_inclusive" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the total number of leaves belonging to each clade on the subtree.
Counts are inclusive (e.g. samples belonging to a clade descended from another clade will count for the ancestor clade as well)
By default, counts across the whole tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.count_haplotypes">
<span class="sig-name descname"><span class="pre">count_haplotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.count_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.count_leaves">
<span class="sig-name descname"><span class="pre">count_leaves</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subroot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.count_leaves" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of leaves descended from the indicated node. By default, counts all leaves on the tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.count_mutation_types">
<span class="sig-name descname"><span class="pre">count_mutation_types</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subroot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.count_mutation_types" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the counts of individual mutation types across the tree. If a subtree root is indicated, it only counts mutations
descended from that node. By default, this counts across the entire tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.depth_first_expansion">
<span class="sig-name descname"><span class="pre">depth_first_expansion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.depth_first_expansion" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a preorder (depth-first) expansion of the tree, starting from the indicated node. 
By default, traverses the whole tree. Set reverse to true to traverse in postorder (reverse depth-first) instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.from_json">
<span class="sig-name descname"><span class="pre">from_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jsonf</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.from_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a mat from a json compatible with the Auspice.us visualization web tool.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.from_newick">
<span class="sig-name descname"><span class="pre">from_newick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nwk</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.from_newick" title="Permalink to this definition">¶</a></dt>
<dd><p>Load from a newick only. The resulting tree will lack mutation information, preventing some functions from being applied.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.from_newick_and_vcf">
<span class="sig-name descname"><span class="pre">from_newick_and_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nwk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.from_newick_and_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Load from a newick and a vcf. The vcf must contain sample entries (genotype columns) for every leaf in the newick.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.from_pb">
<span class="sig-name descname"><span class="pre">from_pb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncondense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.from_pb" title="Permalink to this definition">¶</a></dt>
<dd><p>Load from a protobuf. Includes both tree and mutation information.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_clade">
<span class="sig-name descname"><span class="pre">get_clade</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clade_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_clade" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subtree representing the selected clade.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_clade_samples">
<span class="sig-name descname"><span class="pre">get_clade_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clade_id</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">vector</span><span class="p"><span class="pre">[</span></span><span class="pre">string</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#bte.MATree.get_clade_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Return samples from the selected clade.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_leaves">
<span class="sig-name descname"><span class="pre">get_leaves</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_leaves" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a list of MATNode objects representing each leaf descended from the indicated node. By default, returns all leaves on the tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_leaves_ids">
<span class="sig-name descname"><span class="pre">get_leaves_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_leaves_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of leaf name strings containing all leaves descended from the indicated node. By default, returns all leaves on the tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_mutation">
<span class="sig-name descname"><span class="pre">get_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subtree of samples containing the indicated mutation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_mutation_samples">
<span class="sig-name descname"><span class="pre">get_mutation_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">vector</span><span class="p"><span class="pre">[</span></span><span class="pre">string</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#bte.MATree.get_mutation_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Return samples containing the selected mutation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_node">
<span class="sig-name descname"><span class="pre">get_node</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a MATNode class object representing the indicated node.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_parsimony_score">
<span class="sig-name descname"><span class="pre">get_parsimony_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_parsimony_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the parsimony score of the complete tree.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_random">
<span class="sig-name descname"><span class="pre">get_random</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lca_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_random" title="Permalink to this definition">¶</a></dt>
<dd><p>Select a random subtree of the selected size. Optionally, pass a list of samples to include.
If the list of samples to include is larger than the target size, random samples will be removed from the list.
Set lca_limit to True to limit random selection to below the common ancestor of the current selection.
Selects as many as possible if not enough are available.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_regex">
<span class="sig-name descname"><span class="pre">get_regex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regexstr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.get_regex" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subtree containing all samples matching the regular expression.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.get_regex_samples">
<span class="sig-name descname"><span class="pre">get_regex_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regexstr</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">vector</span><span class="p"><span class="pre">[</span></span><span class="pre">string</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#bte.MATree.get_regex_samples" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.ladderize">
<span class="sig-name descname"><span class="pre">ladderize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.ladderize" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the branches of the tree according to the size of each partition.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.list_clades">
<span class="sig-name descname"><span class="pre">list_clades</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.list_clades" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a set of all valid clade annotations in the tree that can be used with get_clade and other functions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.mutation_set">
<span class="sig-name descname"><span class="pre">mutation_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.mutation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complete set of mutations the indicated node has with respect to the reference.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.reverse_strand">
<span class="sig-name descname"><span class="pre">reverse_strand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">29903</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.reverse_strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Inverts the tree representation of mutations such that all mutations are with respect to the reverse strand of the reference.
All bases are complemented and indeces are reversed. The tree structure itself and parsimony scores are unaffected.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.rsearch">
<span class="sig-name descname"><span class="pre">rsearch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.rsearch" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of MATNode objects representing the ancestors of the indicated node back to the root in order from the node to the root.
Include_self will include the indicated node on the path. Reverse will yield the path in reverse order (root to node).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.save_pb">
<span class="sig-name descname"><span class="pre">save_pb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condense</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.save_pb" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.simple_parsimony">
<span class="sig-name descname"><span class="pre">simple_parsimony</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_assignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.simple_parsimony" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is an implementation of the small parsimony problem (Fitch algorithm) for a single set of states.
It takes as input a dictionary mapping leaf names to character states and returns a dictionary mapping both leaf and internal node names to inferred character states.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.subtree">
<span class="sig-name descname"><span class="pre">subtree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.subtree" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subtree containing all samples in the input list.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.write_json">
<span class="sig-name descname"><span class="pre">write_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jsonf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Tree'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metafiles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.write_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Writ a json compatible with the Auspice.us visualization web tool containing the indicated samples.
Default behavior includes the whole tree.
You can optionally pass a tsv or csv file or a list of tsv and csv files
containing categorical metadata to decorate the json with (one sample per row).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.write_newick">
<span class="sig-name descname"><span class="pre">write_newick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subroot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_internal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_branch_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retain_original_branch_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncondense_leaves</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.write_newick" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bte.MATree.write_vcf">
<span class="sig-name descname"><span class="pre">write_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_genotypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bte.MATree.write_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a vcf representing the chosen samples to the indicated file. By default, writes a vcf including all samples.</p>
</dd></dl>

</dd></dl>

</section>


          </div>
          
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Jakob McBroome.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>